Scripts and Documentation on how to run Pshtt on multiple GCE Instances

amirian28 opened the pull request:
It is possible that an individual would like to run Pshtt at a large scale (1 million or more) and doesn't have internal infrastructure to do this. For about $100-200, a user can use these scripts to run Pshtt on Google Cloud Engine on 1 Million sites.

The following PR has an array of scripts that will help a user run pshtt at a large scale using GCE.

konklone commented on gce-scripts/README.md:
```diff
@@ -0,0 +1,205 @@
+        roots.pem.
+    *   domains must have the schema stripped of them and no trailing '/', such
+        as:
+        *   domain.tld/path/to/page
```
`domain.tld/path/to/page` isn't a supported thing to do with `pshtt`, and unfortunately right now will even screw up the base domain calculator.

konklone commented on gce-scripts/packages_to_install.sh:
```diff
@@ -0,0 +1,51 @@
+apt-get -y update -qq
+echo $? ' ERROR CODE'
+echo 'PYTHON PIP'
+apt-get -y install python-pip -qq
```
Is `python-pip` required in addition to `python3-pip` below? I don't think we have any Python 2 dependencies anymore.

konklone commented on gce-scripts/packages_to_install.sh:
```diff
@@ -0,0 +1,51 @@
+apt-get -y install libffi6 libffi-dev -qq
+echo $? ' ERROR CODE'
+echo 'LIBSSL'
+apt-get -y install build-essential libssl-dev libffi-dev python-dev python3-dev -qq
```
Same question here with `python-dev` and `python3-dev`.

konklone commented on gce-scripts/running_script.sh:
```diff
@@ -0,0 +1,11 @@
+#!/bin/bash
+
+# Runs pshtt with a 10 second time, "Google-Transparency-Report" as the user
```
This cites `"Google-Transparency-Report"` as the user agent, but this may not be intended, and I don't see this string mentioned anywhere else in the PR.

konklone replied:
Also I think this is meant to be `10 second timeout` rather than `10 second time`?

konklone commented:
LGTM - @h-m-f-t, you a :+1: to merge?

IanLee1521 commented on gce-scripts/combine_shards.py:
```diff
@@ -0,0 +1,29 @@
+        exit(1)
+    # Master file is the file with the list of filenames to intake.
+    # Fileception.
+    master_file = sys.argv[1]
```
May not be, but is it necessary to do any error handling here? Either check the file exists or catch the exception?

IanLee1521 commented on gce-scripts/packages_to_install.sh:
```diff
@@ -0,0 +1,48 @@
+# Installs all the necessary packages for pshtt to run.
+# Logs which package it is installing as well as it's success (0) or failure
+# (1).
+echo 'UPDATE'
```
Rather than all the `echo ...` 's, would it be sufficient to make use of `set -x` bash command?

IanLee1521 commented on gce-scripts/packages_to_install.sh:
```diff
@@ -0,0 +1,48 @@
+#!/bin/bash
+
+# Installs all the necessary packages for pshtt to run.
+# Logs which package it is installing as well as it's success (0) or failure
```
Is failure OK in this case or should the script error out if a package install fails? If so, perhaps it would be simpler to use `set -e` at the top of the file?

IanLee1521 commented on gce-scripts/running_script.sh:
```diff
@@ -0,0 +1,10 @@
+# ./running_script.sh test_file.csv
+# output files: test_file.csv.json, time_test_file.csv.txt
+
+input_file=$1
```
Might be worth doing a check to make sure there was an argument provided?

IanLee1521 commented on gce-scripts/scp_and_setup.sh:
```diff
@@ -0,0 +1,139 @@
+    # Do not do strict host key checking so that you dont have to type "yes" for
+    # each machine.
+    echo 'Uploading packages_to_install.sh'
+    scp -i ~/.ssh/gce_pshtt_key -o "StrictHostKeyChecking no" packages_to_install.sh ubuntu@"${machine}":~/
```
Is there a reason to do `-o "StrictHostKeyChecking no"` here, but not in the other commands? I see there is the comment above, but will it re-prompt at the next call?

IanLee1521 commented on gce-scripts/scp_and_setup.sh:
```diff
@@ -0,0 +1,139 @@
+    # We echo after each command to ensure that it worked. 0 means success.
+    # The Log file is how we can tell if the packages have all been uploaded.
+    echo 'Creating packages log file'
+    ssh -i ~/.ssh/gce_pshtt_key ubuntu@"${machine}" touch package_log_file.txt
```
Would it be preferable to make a file level alias up above to factor out the ssh requirements? Something like `alias _ssh='ssh -i ~/.ssh/gce_pshtt_key -o "StrictHostKeyChecking no" ubuntu@"${machine}"'` which could then simplify this and other lines to: `_ssh touch package_log_file.txt`

IanLee1521 commented on gce-scripts/scp_and_setup.sh:
```diff
@@ -0,0 +1,139 @@
+    ssh -i ~/.ssh/gce_pshtt_key -t ubuntu@"${machine}" git clone https://github.com/dhs-ncats/pshtt.git
+    echo $?
+    echo 'copying data file to pshtt directory'
+    scp -i ~/.ssh/gce_pshtt_key input_files/"${y}" ubuntu@"${machine}":~/pshtt/
```
Similar to the comment above about an `_ssh` alias, maybe there could be a similar command for `alias _scp='scp -i ~/.ssh/gce_pshtt_key -o "StrictHostKeyChecking no" '`

IanLee1521 commented on gce-scripts/run_all_scripts.sh:
```diff
@@ -0,0 +1,49 @@
+
+input_file=$1
+number_of_shards=${2-10}
+output_file_name=${3-shard_}
```
Maybe I'm missing it, but what is `shard_` ? I don't see it set anyplace else...

h-m-f-t commented:
Thanks @amirian28! @dhs-ncats won't give me a few extra Benjamins to actually try this out on GCE, but I'm happy to add this documentation into the repo and I'm glad that you would share it upstream. 🎉

amirian28 commented:
Thanks! And sorry @IanLee1521 for not addressing your comments --- work got unexpectedly hectic and this PR slipped my mind. Those are great comments/suggestions, so I'm hoping to go back and fix them up in the near future. I'm no bash expert, so the comments are appreciated.

IanLee1521 commented:
No worries, happens to us all. :)

As for the comments, I'm no bash expert either, so happy to help out and continue the conversation on the next pull request. :)
